File: sdks/javascript/xagent-partner.ts
/**
 * X-Agent Partner SDK for JavaScript/TypeScript
 *
 * A comprehensive SDK for integrating with X-Agent Partner API.
 *
 * Installation:
 *   npm install xagent-partner-sdk
 *
 * Usage:
 *   import { PartnerClient } from 'xagent-partner-sdk';
 *
 *   const client = new PartnerClient({ apiKey: 'xag_partner_xxx' });
 *   const partner = await client.getPartner('partner_id');
 */

import crypto from 'crypto';

export const VERSION = '1.0.0';

// ============================================================================
// TYPES
// ============================================================================

export interface PartnerClientConfig {
  apiKey: string;
  baseUrl?: string;
  timeout?: number;
  maxRetries?: number;
}

export interface PartnerRegistrationRequest {
  company_name: string;
  contact_email: string;
  contact_name: string;
  company_website?: string;
  description?: string;
  integration_type?: string;
  use_cases?: string[];
  expected_volume?: string;
}

export interface PartnerResponse {
  partner_id: string;
  company_name: string;
  contact_email: string;
  contact_name: string;
  company_website?: string;
  description?: string;
  integration_type: string;
  use_cases: string[];
  status: string;
  created_at: string;
  updated_at: string;
  api_key_prefix?: string;
  webhook_url?: string;
  monthly_requests: number;
  monthly_limit: number;
}

export interface APIKeyRequest {
  name: string;
  expires_in_days?: number;
  rate_limit_rpm?: number;
  rate_limit_rph?: number;
  ip_whitelist?: string[];
  scopes?: string[];
}

export interface APIKeyResponse {
  key_id: string;
  key?: string;
  key_prefix: string;
  name: string;
  partner_id: string;
  created_at: string;
  expires_at?: string;
  rate_limit_rpm: number;
  rate_limit_rph: number;
  ip_whitelist?: string[];
  scopes: string[];
  status: string;
}

export interface WebhookRequest {
  event_type: string;
  url: string;
  active?: boolean;
  retry_policy?: Record<string, any>;
}

export interface WebhookResponse {
  webhook_id: string;
  partner_id: string;
  event_type: string;
  url: string;
  active: boolean;
  created_at: string;
  last_triggered_at?: string;
  retry_policy?: Record<string, any>;
  delivery_count: number;
  failure_count: number;
}

export interface UsageResponse {
  partner_id: string;
  period_start: string;
  period_end: string;
  total_requests: number;
  successful_requests: number;
  failed_requests: number;
  average_response_time_ms: number;
  api_calls_by_endpoint: Record<string, number>;
  errors_by_type: Record<string, number>;
  bandwidth_used_mb: number;
}

export interface QuotaResponse {
  partner_id: string;
  monthly_limit: number;
  monthly_used: number;
  monthly_remaining: number;
  reset_date: string;
  quota_exceeded: boolean;
}

export interface SupportTicketRequest {
  subject: string;
  description: string;
  priority?: string;
  category?: string;
  attachments?: string[];
}

export interface SupportTicketResponse {
  ticket_id: string;
  partner_id: string;
  subject: string;
  description: string;
  priority: string;
  category: string;
  status: string;
  created_at: string;
  updated_at: string;
  assigned_to?: string;
  resolution_notes?: string;
}

// ============================================================================
// ERRORS
// ============================================================================

export class PartnerAPIError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public response?: Record<string, any>,
  ) {
    super(message);
    this.name = 'PartnerAPIError';
  }
}

export class PartnerAuthError extends PartnerAPIError {
  constructor(message: string, response?: Record<string, any>) {
    super(message, 401, response);
    this.name = 'PartnerAuthError';
  }
}

export class PartnerNotFoundError extends PartnerAPIError {
  constructor(message: string, response?: Record<string, any>) {
    super(message, 404, response);
    this.name = 'PartnerNotFoundError';
  }
}

export class PartnerRateLimitError extends PartnerAPIError {
  constructor(message: string, response?: Record<string, any>) {
    super(message, 429, response);
    this.name = 'PartnerRateLimitError';
  }
}

// ============================================================================
// CLIENT
// ============================================================================

export class PartnerClient {
  private apiKey: string;
  private baseUrl: string;
  private timeout: number;
  private maxRetries: number;

  constructor(config: PartnerClientConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = (config.baseUrl || 'https://api.x-agent.io').replace(/\/$/, '');
    this.timeout = config.timeout || 30000;
    this.maxRetries = config.maxRetries || 3;
  }

  private getHeaders(): Record<string, string> {
    return {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      'User-Agent': `xagent-partner-sdk/${VERSION}`,
    };
  }

  private async makeRequest<T>(
    method: string,
    endpoint: string,
    options?: {
      body?: Record<string, any>;
      params?: Record<string, any>;
    },
  ): Promise<T> {
    let url = `${this.baseUrl}${endpoint}`;

    if (options?.params) {
      const searchParams = new URLSearchParams();
      for (const [key, value] of Object.entries(options.params)) {
        if (value !== undefined && value !== null) {
          searchParams.append(key, String(value));
        }
      }
      const queryString = searchParams.toString();
      if (queryString) {
        url += `?${queryString}`;
      }
    }

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        const response = await fetch(url, {
          method,
          headers: this.getHeaders(),
          body: options?.body ? JSON.stringify(options.body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        // Handle rate limiting
        if (response.status === 429) {
          const retryAfter = parseInt(response.headers.get('Retry-After') || '60', 10);
          if (attempt < this.maxRetries - 1) {
            await this.sleep(retryAfter * 1000);
            continue;
          }
          const data = await response.json();
          throw new PartnerRateLimitError('Rate limit exceeded', data);
        }

        // Handle server errors
        if (response.status >= 500) {
          if (attempt < this.maxRetries - 1) {
            const waitTime = Math.pow(2, attempt) * 1000;
            await this.sleep(waitTime);
            continue;
          }
        }

        // Handle client errors
        if (response.status === 401) {
          const data = await response.json();
          throw new PartnerAuthError('Unauthorized', data);
        }

        if (response.status === 404) {
          const data = await response.json();
          throw new PartnerNotFoundError('Resource not found', data);
        }

        if (response.status >= 400) {
          const data = await response.json();
          throw new PartnerAPIError(
            data.error?.message || 'API error',
            response.status,
            data,
          );
        }

        return await response.json();
      } catch (error) {
        if (error instanceof PartnerAPIError) {
          throw error;
        }

        if (attempt === this.maxRetries - 1) {
          throw new PartnerAPIError(`Request failed: ${error}`);
        }

        const waitTime = Math.pow(2, attempt) * 1000;
        await this.sleep(waitTime);
      }
    }

    throw new PartnerAPIError('Max retries exceeded');
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  // ========================================================================
  // PARTNER MANAGEMENT
  // ========================================================================

  async registerPartner(request: PartnerRegistrationRequest): Promise<PartnerResponse> {
    return this.makeRequest<PartnerResponse>('POST', '/api/v1/partners/register', {
      body: request,
    });
  }

  async getPartner(partnerId: string): Promise<PartnerResponse> {
    return this.makeRequest<PartnerResponse>('GET', `/api/v1/partners/${partnerId}`);
  }

  async listPartners(options?: {
    statusFilter?: string;
    integrationType?: string;
    skip?: number;
    limit?: number;
  }): Promise<PartnerResponse[]> {
    const response = await this.makeRequest<PartnerResponse[] | { partners: PartnerResponse[] }>(
      'GET',
      '/api/v1/partners',
      {
        params: {
          status_filter: options?.statusFilter,
          integration_type: options?.integrationType,
          skip: options?.skip,
          limit: options?.limit,
        },
      },
    );
    return Array.isArray(response) ? response : response.partners || [];
  }

  async updatePartner(partnerId: string, updates: Record<string, any>): Promise<PartnerResponse> {
    return this.makeRequest<PartnerResponse>('PATCH', `/api/v1/partners/${partnerId}`, {
      body: updates,
    });
  }

  async approvePartner(partnerId: string): Promise<PartnerResponse> {
    return this.makeRequest<PartnerResponse>('POST', `/api/v1/partners/${partnerId}/approve`);
  }

  async suspendPartner(partnerId: string, reason: string): Promise<PartnerResponse> {
    return this.makeRequest<PartnerResponse>('POST', `/api/v1/partners/${partnerId}/suspend`, {
      params: { reason },
    });
  }

  // ========================================================================
  // API KEY MANAGEMENT
  // ========================================================================

  async createAPIKey(partnerId: string, request: APIKeyRequest): Promise<APIKeyResponse> {
    return this.makeRequest<APIKeyResponse>('POST', `/api/v1/partners/${partnerId}/api-keys`, {
      body: request,
    });
  }

  async listAPIKeys(
    partnerId: string,
    statusFilter?: string,
  ): Promise<APIKeyResponse[]> {
    const response = await this.makeRequest<APIKeyResponse[] | { keys: APIKeyResponse[] }>(
      'GET',
      `/api/v1/partners/${partnerId}/api-keys`,
      {
        params: { status_filter: statusFilter },
      },
    );
    return Array.isArray(response) ? response : response.keys || [];
  }

  async rotateAPIKey(partnerId: string, keyId: string): Promise<APIKeyResponse> {
    return this.makeRequest<APIKeyResponse>(
      'POST',
      `/api/v1/partners/${partnerId}/api-keys/${keyId}/rotate`,
    );
  }

  async revokeAPIKey(partnerId: string, keyId: string): Promise<void> {
    await this.makeRequest<void>('DELETE', `/api/v1/partners/${partnerId}/api-keys/${keyId}`);
  }

  // ========================================================================
  // WEBHOOK MANAGEMENT
  // ========================================================================

  async registerWebhook(partnerId: string, request: WebhookRequest): Promise<WebhookResponse> {
    return this.makeRequest<WebhookResponse>('POST', `/api/v1/partners/${partnerId}/webhooks`, {
      body: request,
    });
  }

  async listWebhooks(partnerId: string): Promise<WebhookResponse[]> {
    const response = await this.makeRequest<WebhookResponse[] | { webhooks: WebhookResponse[] }>(
      'GET',
      `/api/v1/partners/${partnerId}/webhooks`,
    );
    return Array.isArray(response) ? response : response.webhooks || [];
  }

  async updateWebhook(
    partnerId: string,
    webhookId: string,
    updates: Record<string, any>,
  ): Promise<WebhookResponse> {
    return this.makeRequest<WebhookResponse>(
      'PATCH',
      `/api/v1/partners/${partnerId}/webhooks/${webhookId}`,
      { body: updates },
    );
  }

  async deleteWebhook(partnerId: string, webhookId: string): Promise<void> {
    await this.makeRequest<void>('DELETE', `/api/v1/partners/${partnerId}/webhooks/${webhookId}`);
  }

  async testWebhook(partnerId: string, webhookId: string): Promise<Record<string, any>> {
    return this.makeRequest<Record<string, any>>(
      'POST',
      `/api/v1/partners/${partnerId}/webhooks/${webhookId}/test`,
    );
  }

  // ========================================================================
  // USAGE & ANALYTICS
  // ========================================================================

  async getUsage(
    partnerId: string,
    options?: {
      period?: string;
      startDate?: string;
      endDate?: string;
    },
  ): Promise<UsageResponse> {
    return this.makeRequest<UsageResponse>('GET', `/api/v1/partners/${partnerId}/usage`, {
      params: {
        period: options?.period,
        start_date: options?.startDate,
        end_date: options?.endDate,
      },
    });
  }

  async getDailyUsage(
    partnerId: string,
    days?: number,
  ): Promise<Record<string, any>[]> {
    const response = await this.makeRequest<
      Record<string, any>[] | { daily_usage: Record<string, any>[] }
    >('GET', `/api/v1/partners/${partnerId}/usage/daily`, {
      params: { days },
    });
    return Array.isArray(response) ? response : response.daily_usage || [];
  }

  async getQuota(partnerId: string): Promise<QuotaResponse> {
    return this.makeRequest<QuotaResponse>('GET', `/api/v1/partners/${partnerId}/quota`);
  }

  // ========================================================================
  // SUPPORT TICKETS
  // ========================================================================

  async createSupportTicket(
    partnerId: string,
    request: SupportTicketRequest,
  ): Promise<SupportTicketResponse> {
    return this.makeRequest<SupportTicketResponse>(
      'POST',
      `/api/v1/partners/${partnerId}/support/tickets`,
      { body: request },
    );
  }

  async listSupportTickets(
    partnerId: string,
    options?: {
      statusFilter?: string;
      priorityFilter?: string;
      skip?: number;
      limit?: number;
    },
  ): Promise<SupportTicketResponse[]> {
    const response = await this.makeRequest<
      SupportTicketResponse[] | { tickets: SupportTicketResponse[] }
    >('GET', `/api/v1/partners/${partnerId}/support/tickets`, {
      params: {
        status_filter: options?.statusFilter,
        priority_filter: options?.priorityFilter,
        skip: options?.skip,
        limit: options?.limit,
      },
    });
    return Array.isArray(response) ? response : response.tickets || [];
  }

  async getSupportTicket(partnerId: string, ticketId: string): Promise<SupportTicketResponse> {
    return this.makeRequest<SupportTicketResponse>(
      'GET',
      `/api/v1/partners/${partnerId}/support/tickets/${ticketId}`,
    );
  }

  async updateSupportTicket(
    partnerId: string,
    ticketId: string,
    updates: Record<string, any>,
  ): Promise<SupportTicketResponse> {
    return this.makeRequest<SupportTicketResponse>(
      'PATCH',
      `/api/v1/partners/${partnerId}/support/tickets/${ticketId}`,
      { body: updates },
    );
  }

  // ========================================================================
  // DASHBOARD & HEALTH
  // ========================================================================

  async getDashboard(partnerId: string): Promise<Record<string, any>> {
    return this.makeRequest<Record<string, any>>('GET', `/api/v1/partners/${partnerId}/dashboard`);
  }

  async getHealth(partnerId: string): Promise<Record<string, any>> {
    return this.makeRequest<Record<string, any>>('GET', `/api/v1/partners/${partnerId}/health`);
  }

  // ========================================================================
  // WEBHOOK VERIFICATION
  // ========================================================================

  static verifyWebhookSignature(
    requestBody: string | Buffer,
    signature: string,
    secret: string,
  ): boolean {
    const body = typeof requestBody === 'string' ? requestBody : requestBody.toString();
    const expectedSignature = crypto
      .createHmac('sha256', secret)
      .update(body)
      .digest('hex');
    return crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(expectedSignature));
  }
}

// ============================================================================
// EXAMPLE USAGE
// ============================================================================

/*
async function example() {
  const client = new PartnerClient({ apiKey: 'xag_partner_xxx' });

  try {
    // Register partner
    const partner = await client.registerPartner({
      company_name: 'Acme Corp',
      contact_email: 'contact@acme.com',
      contact_name: 'John Doe',
      integration_type: 'standard',
    });
    console.log('Partner registered:', partner.partner_id);

    // Create API key
    const apiKey = await client.createAPIKey(partner.partner_id, {
      name: 'Production Key',
      expires_in_days: 365,
    });
    console.log('API key created:', apiKey.key_prefix);

    // Register webhook
    const webhook = await client.registerWebhook(partner.partner_id, {
      event_type: 'partner.api_key.created',
      url: 'https://acme.com/webhooks/xagent',
    });
    console.log('Webhook registered:', webhook.webhook_id);

    // Get usage
    const usage = await client.getUsage(partner.partner_id);
    console.log('Usage:', usage.total_requests, 'requests');
  } catch (error) {
    console.error('Error:', error);
  }
}
*/

export default PartnerClient;

File: sdks/python/xagent_partner.py
"""
X-Agent Partner SDK for Python.

A comprehensive SDK for integrating with X-Agent Partner API.

Installation (local build, not published to PyPI):
    pip install ./sdks/python        # from repository root
    # or: python -m pip wheel ./sdks/python --no-deps -w dist

Usage:
    from xagent_partner import PartnerClient

    client = PartnerClient(api_key="xag_partner_xxx")
    partner = client.get_partner("partner_id")
"""

from __future__ import annotations

import hashlib
import hmac
import json
import logging
from datetime import datetime, timedelta
from typing import Any, Optional
from urllib.parse import urljoin

import httpx

logger = logging.getLogger(__name__)

__version__ = "0.2.0-alpha"  # 版本单一事实源: pyproject.toml 通过 [tool.setuptools.dynamic] 动态读取本值
__all__ = [
    "PartnerClient",
    "PartnerAPIError",
    "PartnerAuthError",
    "PartnerNotFoundError",
    "PartnerRateLimitError",
]


class PartnerAPIError(Exception):
    """Base exception for Partner API errors."""

    def __init__(self, message: str, status_code: Optional[int] = None, response: Optional[dict] = None):
        self.message = message
        self.status_code = status_code
        self.response = response
        super().__init__(message)


class PartnerAuthError(PartnerAPIError):
    """Authentication error."""

    pass


class PartnerNotFoundError(PartnerAPIError):
    """Resource not found error."""

    pass


class PartnerRateLimitError(PartnerAPIError):
    """Rate limit exceeded error."""

    pass


class PartnerClient:
    """
    X-Agent Partner API Client.

    Provides methods for managing partners, API keys, webhooks, and usage.

    Args:
        api_key: Partner API key
        base_url: Base URL for API (default: https://api.x-agent.io)
        timeout: Request timeout in seconds (default: 30)
        max_retries: Maximum number of retries (default: 3)
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.x-agent.io",
        timeout: int = 30,
        max_retries: int = 3,
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries
        self._client = httpx.Client(
            timeout=timeout,
            headers=self._get_headers(),
        )

    def _get_headers(self) -> dict[str, str]:
        """Get default request headers."""
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "User-Agent": f"xagent-partner-sdk/{__version__}",
        }

    def _make_request(
        self,
        method: str,
        endpoint: str,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Make HTTP request with retry logic."""
        url = urljoin(self.base_url, endpoint)

        for attempt in range(self.max_retries):
            try:
                response = self._client.request(method, url, **kwargs)

                # Handle rate limiting
                if response.status_code == 429:
                    retry_after = int(response.headers.get("Retry-After", 60))
                    if attempt < self.max_retries - 1:
                        import time

                        time.sleep(retry_after)
                        continue
                    raise PartnerRateLimitError(
                        "Rate limit exceeded",
                        status_code=429,
                        response=response.json(),
                    )

                # Handle server errors
                if response.status_code >= 500:
                    if attempt < self.max_retries - 1:
                        import time

                        wait_time = 2 ** attempt
                        time.sleep(wait_time)
                        continue

                # Handle client errors
                if response.status_code == 401:
                    error_data = response.json()
                    raise PartnerAuthError(
                        error_data.get("detail") or "Unauthorized",
                        status_code=401,
                        response=error_data,
                    )

                if response.status_code == 404:
                    error_data = response.json()
                    raise PartnerNotFoundError(
                        error_data.get("detail") or "Resource not found",
                        status_code=404,
                        response=error_data,
                    )

                if response.status_code >= 400:
                    error_data = response.json()
                    message = (
                        error_data.get("error", {}).get("message")
                        or error_data.get("detail")  # FastAPI HTTPException shape
                        or "API error"
                    )
                    if not isinstance(message, str):
                        message = str(message)
                    raise PartnerAPIError(
                        message,
                        status_code=response.status_code,
                        response=error_data,
                    )

                # No-content responses (e.g., 204 from DELETE endpoints)
                if response.status_code == 204 or not response.content:
                    return {}

                return response.json()

            except httpx.RequestError as e:
                if attempt == self.max_retries - 1:
                    raise PartnerAPIError(f"Request failed: {e}")
                import time

                time.sleep(2 ** attempt)

        raise PartnerAPIError("Max retries exceeded")

    def close(self) -> None:
        """Close the HTTP client."""
        self._client.close()

    def __enter__(self) -> PartnerClient:
        """Context manager entry."""
        return self

    def __exit__(self, *args: Any) -> None:
        """Context manager exit."""
        self.close()

    # ========================================================================
    # PARTNER MANAGEMENT
    # ========================================================================

    def register_partner(
        self,
        company_name: str,
        contact_email: str,
        contact_name: str,
        company_website: Optional[str] = None,
        description: Optional[str] = None,
        integration_type: str = "standard",
        use_cases: Optional[list[str]] = None,
        expected_volume: Optional[str] = None,
    ) -> dict[str, Any]:
        """Register a new partner."""
        payload = {
            "company_name": company_name,
            "contact_email": contact_email,
            "contact_name": contact_name,
            "company_website": company_website,
            "description": description,
            "integration_type": integration_type,
            "use_cases": use_cases or [],
            "expected_volume": expected_volume,
        }
        return self._make_request("POST", "/api/v1/partners/register", json=payload)

    def get_partner(self, partner_id: str) -> dict[str, Any]:
        """Get partner information."""
        return self._make_request("GET", f"/api/v1/partners/{partner_id}")

    def list_partners(
        self,
        status_filter: Optional[str] = None,
        integration_type: Optional[str] = None,
        skip: int = 0,
        limit: int = 50,
    ) -> list[dict[str, Any]]:
        """List all partners."""
        params = {
            "skip": skip,
            "limit": limit,
        }
        if status_filter:
            params["status_filter"] = status_filter
        if integration_type:
            params["integration_type"] = integration_type

        response = self._make_request("GET", "/api/v1/partners", params=params)
        return response if isinstance(response, list) else response.get("partners", [])

    def update_partner(self, partner_id: str, **updates: Any) -> dict[str, Any]:
        """Update partner information."""
        return self._make_request("PATCH", f"/api/v1/partners/{partner_id}", json=updates)

    def approve_partner(self, partner_id: str) -> dict[str, Any]:
        """Approve a pending partner."""
        return self._make_request("POST", f"/api/v1/partners/{partner_id}/approve")

    def suspend_partner(self, partner_id: str, reason: str) -> dict[str, Any]:
        """Suspend a partner."""
        return self._make_request(
            "POST",
            f"/api/v1/partners/{partner_id}/suspend",
            params={"reason": reason},
        )

    # ========================================================================
    # API KEY MANAGEMENT
    # ========================================================================

    def create_api_key(
        self,
        partner_id: str,
        name: str,
        expires_in_days: int = 365,
        rate_limit_rpm: int = 1000,
        rate_limit_rph: int = 50000,
        ip_whitelist: Optional[list[str]] = None,
        scopes: Optional[list[str]] = None,
    ) -> dict[str, Any]:
        """Create a new API key for partner."""
        payload = {
            "name": name,
            "expires_in_days": expires_in_days,
            "rate_limit_rpm": rate_limit_rpm,
            "rate_limit_rph": rate_limit_rph,
            "ip_whitelist": ip_whitelist,
            "scopes": scopes or [],
        }
        return self._make_request(
            "POST",
            f"/api/v1/partners/{partner_id}/api-keys",
            json=payload,
        )

    def list_api_keys(
        self,
        partner_id: str,
        status_filter: Optional[str] = None,
    ) -> list[dict[str, Any]]:
        """List API keys for partner."""
        params = {}
        if status_filter:
            params["status_filter"] = status_filter

        response = self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/api-keys",
            params=params,
        )
        return response if isinstance(response, list) else response.get("keys", [])

    def rotate_api_key(self, partner_id: str, key_id: str) -> dict[str, Any]:
        """Rotate an API key."""
        return self._make_request(
            "POST",
            f"/api/v1/partners/{partner_id}/api-keys/{key_id}/rotate",
        )

    def revoke_api_key(self, partner_id: str, key_id: str) -> None:
        """Revoke an API key."""
        self._make_request(
            "DELETE",
            f"/api/v1/partners/{partner_id}/api-keys/{key_id}",
        )

    # ========================================================================
    # WEBHOOK MANAGEMENT
    # ========================================================================

    def register_webhook(
        self,
        partner_id: str,
        event_type: str,
        url: str,
        active: bool = True,
        retry_policy: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Register a webhook."""
        payload = {
            "event_type": event_type,
            "url": url,
            "active": active,
            "retry_policy": retry_policy,
        }
        return self._make_request(
            "POST",
            f"/api/v1/partners/{partner_id}/webhooks",
            json=payload,
        )

    def list_webhooks(self, partner_id: str) -> list[dict[str, Any]]:
        """List webhooks for partner."""
        response = self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/webhooks",
        )
        return response if isinstance(response, list) else response.get("webhooks", [])

    def update_webhook(
        self,
        partner_id: str,
        webhook_id: str,
        **updates: Any,
    ) -> dict[str, Any]:
        """Update webhook configuration."""
        return self._make_request(
            "PATCH",
            f"/api/v1/partners/{partner_id}/webhooks/{webhook_id}",
            json=updates,
        )

    def delete_webhook(self, partner_id: str, webhook_id: str) -> None:
        """Delete a webhook."""
        self._make_request(
            "DELETE",
            f"/api/v1/partners/{partner_id}/webhooks/{webhook_id}",
        )

    def test_webhook(self, partner_id: str, webhook_id: str) -> dict[str, Any]:
        """Send a test event to webhook."""
        return self._make_request(
            "POST",
            f"/api/v1/partners/{partner_id}/webhooks/{webhook_id}/test",
        )

    # ========================================================================
    # USAGE & ANALYTICS
    # ========================================================================

    def get_usage(
        self,
        partner_id: str,
        period: str = "current_month",
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
    ) -> dict[str, Any]:
        """Get partner usage statistics."""
        params = {"period": period}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date

        return self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/usage",
            params=params,
        )

    def get_daily_usage(
        self,
        partner_id: str,
        days: int = 30,
    ) -> list[dict[str, Any]]:
        """Get daily usage breakdown."""
        response = self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/usage/daily",
            params={"days": days},
        )
        return response if isinstance(response, list) else response.get("daily_usage", [])

    def get_quota(self, partner_id: str) -> dict[str, Any]:
        """Get partner quota and usage."""
        return self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/quota",
        )

    # ========================================================================
    # SUPPORT TICKETS
    # ========================================================================

    def create_support_ticket(
        self,
        partner_id: str,
        subject: str,
        description: str,
        priority: str = "normal",
        category: str = "general",
        attachments: Optional[list[str]] = None,
    ) -> dict[str, Any]:
        """Create a support ticket."""
        payload = {
            "subject": subject,
            "description": description,
            "priority": priority,
            "category": category,
            "attachments": attachments,
        }
        return self._make_request(
            "POST",
            f"/api/v1/partners/{partner_id}/support/tickets",
            json=payload,
        )

    def list_support_tickets(
        self,
        partner_id: str,
        status_filter: Optional[str] = None,
        priority_filter: Optional[str] = None,
        skip: int = 0,
        limit: int = 50,
    ) -> list[dict[str, Any]]:
        """List support tickets."""
        params = {"skip": skip, "limit": limit}
        if status_filter:
            params["status_filter"] = status_filter
        if priority_filter:
            params["priority_filter"] = priority_filter

        response = self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/support/tickets",
            params=params,
        )
        return response if isinstance(response, list) else response.get("tickets", [])

    def get_support_ticket(self, partner_id: str, ticket_id: str) -> dict[str, Any]:
        """Get support ticket details."""
        return self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/support/tickets/{ticket_id}",
        )

    def update_support_ticket(
        self,
        partner_id: str,
        ticket_id: str,
        **updates: Any,
    ) -> dict[str, Any]:
        """Update support ticket."""
        return self._make_request(
            "PATCH",
            f"/api/v1/partners/{partner_id}/support/tickets/{ticket_id}",
            json=updates,
        )

    # ========================================================================
    # DASHBOARD & HEALTH
    # ========================================================================

    def get_dashboard(self, partner_id: str) -> dict[str, Any]:
        """Get partner dashboard data."""
        return self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/dashboard",
        )

    def get_health(self, partner_id: str) -> dict[str, Any]:
        """Get partner integration health status."""
        return self._make_request(
            "GET",
            f"/api/v1/partners/{partner_id}/health",
        )

    # ========================================================================
    # WEBHOOK VERIFICATION
    # ========================================================================

    @staticmethod
    def verify_webhook_signature(
        request_body: bytes,
        signature: str,
        secret: str,
    ) -> bool:
        """
        Verify webhook signature.

        Args:
            request_body: Raw request body bytes
            signature: Signature from X-Webhook-Signature header
            secret: Webhook secret

        Returns:
            True if signature is valid, False otherwise
        """
        expected_signature = hmac.new(
            secret.encode(),
            request_body,
            hashlib.sha256,
        ).hexdigest()
        return hmac.compare_digest(signature, expected_signature)


# ============================================================================
# EXAMPLE USAGE
# ============================================================================

if __name__ == "__main__":
    # Example: Register partner and create API key
    with PartnerClient(api_key="xag_partner_xxx") as client:
        # Register partner
        partner = client.register_partner(
            company_name="Acme Corp",
            contact_email="contact@acme.com",
            contact_name="John Doe",
            integration_type="standard",
        )
        print(f"Partner registered: {partner['partner_id']}")

        # Create API key
        api_key = client.create_api_key(
            partner_id=partner["partner_id"],
            name="Production Key",
            expires_in_days=365,
        )
        print(f"API key created: {api_key['key_prefix']}")

        # Register webhook
        webhook = client.register_webhook(
            partner_id=partner["partner_id"],
            event_type="partner.api_key.created",
            url="https://acme.com/webhooks/xagent",
        )
        print(f"Webhook registered: {webhook['webhook_id']}")

        # Get usage
        usage = client.get_usage(partner_id=partner["partner_id"])
        print(f"Usage: {usage['total_requests']} requests")
